### PR TITLE
HIVE-27327 : Iceberg basic stats: Incorrect row count in snapshot sum…

### PR DESCRIPTION
**Pull request by simhadri-g**
…mary leading to unoptimized plans

### CODE REVIEW

File: iceberg/iceberg-handler/src/test/queries/positive/row_count.q [hunk 1]
```diff
@@ -0,0 +1,42 @@
+drop table llap_orders;
```
Comment (InvisibleProgrammer) [inline iceberg/iceberg-handler/src/test/queries/positive/row_count.q:1]: I'm not 100% sure but maybe there can be an edge case when the table is not cleaned up properly. In that case, I would consider using `DROP TABLE IF EXISTS...`

File: iceberg/iceberg-handler/src/test/results/positive/row_count.q.out [hunk 2]
```diff
@@ -0,0 +1,302 @@
+PREHOOK: query: drop table llap_orders
```
Comment (InvisibleProgrammer) [inline iceberg/iceberg-handler/src/test/results/positive/row_count.q.out:1]: Please mask out the values that can change frequently, like `current-snapshot-id`, `added-files-size`, etc...

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergStorageHandler.java [hunk 3]
```diff
@@ -346,7 +346,16 @@ public Map<String, String> getBasicStatistics(Partish partish) {
               stats.put(StatsSetupConst.NUM_FILES, summary.get(SnapshotSummary.TOTAL_DATA_FILES_PROP));
             }
             if (summary.containsKey(SnapshotSummary.TOTAL_RECORDS_PROP)) {
-              stats.put(StatsSetupConst.ROW_COUNT, summary.get(SnapshotSummary.TOTAL_RECORDS_PROP));
+              long totalRecords = Long.parseLong(summary.get(SnapshotSummary.TOTAL_RECORDS_PROP));
+              if (summary.containsKey(SnapshotSummary.TOTAL_EQ_DELETES_PROP) &&
```
Comment (InvisibleProgrammer) [inline iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergStorageHandler.java:350]: What if onlye one of `TOTAL_EQ_DELETES_PROP` and `TOTAL_POS_DELETES_PROP` persists?

**Review by InvisibleProgrammer — COMMENTED** [review]
If I understood it correctly, it runs when we query the statistics, not when we write them. I wonder, it is possible to fix it on writes?

**Review by InvisibleProgrammer — COMMENTED** [review]
If I understood it correctly, it runs when we query the statistics, not when we write them. I wonder, it is possible to fix it on writes?

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergStorageHandler.java [hunk 4]
```diff
@@ -346,7 +346,16 @@ public Map<String, String> getBasicStatistics(Partish partish) {
...
+                  summary.containsKey(SnapshotSummary.TOTAL_POS_DELETES_PROP)) {
+                Long actualRecords =
+                    totalRecords - (Long.parseLong(summary.get(SnapshotSummary.TOTAL_EQ_DELETES_PROP)) >
```
Comment (aturoczy) [inline iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergStorageHandler.java:353]: It is a code smell a bit imho.

When you are using conditional operator your eyes need to parse a variable that you put it into 2 methods. So to read / debug the value you need to have an understanding what the summary.get returns and after the Long.prase (last is easier for sure.  Maybe It is fully OK and you can choose to not accept this opinion but I think it is nicer if you create a variable for those values.

Comment (aturoczy) [reply iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergStorageHandler.java:353]: Btw why this getBasicStatistics needs switch case?

**Review by aturoczy — CHANGES_REQUESTED** [review]
Please reject if you not agree with my findings

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergStorageHandler.java [hunk 5]
```diff
@@ -346,7 +346,16 @@ public Map<String, String> getBasicStatistics(Partish partish) {
               stats.put(StatsSetupConst.NUM_FILES, summary.get(SnapshotSummary.TOTAL_DATA_FILES_PROP));
             }
             if (summary.containsKey(SnapshotSummary.TOTAL_RECORDS_PROP)) {
-              stats.put(StatsSetupConst.ROW_COUNT, summary.get(SnapshotSummary.TOTAL_RECORDS_PROP));
+              long totalRecords = Long.parseLong(summary.get(SnapshotSummary.TOTAL_RECORDS_PROP));
+              if (summary.containsKey(SnapshotSummary.TOTAL_EQ_DELETES_PROP) &&
+                  summary.containsKey(SnapshotSummary.TOTAL_POS_DELETES_PROP)) {
+                Long actualRecords =
```
Comment (zhangbutao) [inline iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergStorageHandler.java:352]: Just share some my thought.
Not sure if i am understand correctly, the delete file in iceberg is also a special data file, and table scan in actual execution stage also should read all related delete files.

That is to say, the actual execution still requires scanning more data than the explain shows.
So, i am not sure if this PR can be give a optimized plans when iceberg table has both data files and delete files.

File: iceberg/iceberg-handler/src/test/queries/positive/row_count.q (hunk 1 above)
Comment (simhadri-g) [reply iceberg/iceberg-handler/src/test/queries/positive/row_count.q]: Fixed

File: iceberg/iceberg-handler/src/test/results/positive/row_count.q.out (hunk 2 above)
Comment (simhadri-g) [reply iceberg/iceberg-handler/src/test/results/positive/row_count.q.out]: Fixed

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergStorageHandler.java (hunk 4 above)
Comment (simhadri-g) [reply iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergStorageHandler.java]: Fixed. 

getBasicStatistics doesn't need a switch case anymore.
Thanks! :)

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergStorageHandler.java (hunk 5 above)
Comment (deniskuzZ) [reply iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergStorageHandler.java]: count(*) should use stats instead of scanning the whole dataset. 
To be honest I don't really understand the purpose of 'total-records' if it doesn't reflect an accurate row count. insert 100 rows, delete all, 'total-records'=100 ???

File: iceberg/iceberg-handler/src/test/results/positive/vectorized_iceberg_merge_mixed.q.out [hunk 6]
```diff
@@ -400,21 +345,21 @@ STAGE PLANS:
...
-                      outputColumnNames: _col0, _col1, _col2, _col3, _col4, _col5, _col6, _col7, _col8, _col9, _col10, _col11, _col12, _col13, _col14, _col15, _col16, _col17, _col18, _col19, _col20, _col21, _col22, _col23, _col24, _col25, _col26, _col27, _col28, _col29, _col30, _col31, _col32, _col33, _col34, _col35, _col36, _col37, _col38, _col39, _col40, _col41, _col42, _col43, _col44, _col45, _col46, _col47, _col48
+                      outputColumnNames: _col0, _col1, _col2, _col3, _col4, _col5, _col6, _col7, _col8, _col9, _col10, _col11, _col12, _col13, _col14, _col15, _col16, _col17, _col18, _col19, _col20, _col21, _col24, _col25, _col26, _col27, _col28, _col29, _col30, _col31, _col32, _col33, _col34, _col35, _col36, _col37, _col38, _col39, _col40, _col41, _col42, _col43, _col44, _col45, _col46, _col47, _col48, _col49, _col50
                       input vertices:
-                        0 Map 1
-                      Statistics: Num rows: 5 Data size: 4320 Basic stats: COMPLETE Column stats: COMPLETE
+                        1 Map 6
+                      Statistics: Num rows: 7 Data size: 6656 Basic stats: COMPLETE Column stats: COMPLETE
```
Comment (deniskuzZ) [inline iceberg/iceberg-handler/src/test/results/positive/vectorized_iceberg_merge_mixed.q.out:354]: was the row count completely inaccurate before?

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergStorageHandler.java (hunk 5 above)
Comment (zhangbutao) [reply iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergStorageHandler.java]: > count(*) should use stats instead of scanning the whole dataset.

@deniskuzZ Do you mean we can push down min/max to iceberg? I think it's beyond this PR scope and it's not so easy. I find some info: HIVE-27099` select count(*) from table queries all data`, and Spark has push down min/max to iceberg https://github.com/apache/iceberg/pull/6622, but Spark will skip pushdown if including delete files https://github.com/apache/iceberg/pull/6622/files#diff-66bfda4bda6d505fe3de7db3b4d6b7923b3711b00e2801846dd7325edcdbf65eR224

> @zhangbutao do you see any issues with that?

To be honest, i  don't know too much about iceberg statistics at the moment so I can't share any more context. 
Maybe after some time I can tell more or Implement some stats pushdown in Hive. :)

Comment (deniskuzZ) [reply iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergStorageHandler.java]: @zhangbutao, we are already pushing down column stats in puffin format [HIVE-27158](https://issues.apache.org/jira/browse/HIVE-27158), however, we are still relying on Iceberg for basic stats. 
In case of deletes, it becomes invalid. In this PR we are doing a workaround just for positional deletes use-case until it's fixed on the iceberg side.

Comment (zhangbutao) [reply iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergStorageHandler.java]: I see. Just say something else here. I think TrinoDB has a good implementation about puffin stats, maybe we can refer to some designs from TrinoDB. But, I also think lots of stuff need to be done on the iceberg side.
We can keep looking at the evolution about iceberg stats, as this can give a better cbo, pushdown, etc

File: iceberg/iceberg-handler/src/test/results/positive/vectorized_iceberg_merge_mixed.q.out (hunk 6 above)
Comment (simhadri-g) [reply iceberg/iceberg-handler/src/test/results/positive/vectorized_iceberg_merge_mixed.q.out:372]: No.

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergStorageHandler.java [hunk 7]
```diff
@@ -337,32 +337,39 @@ public Map<String, String> getBasicStatistics(Partish partish) {
...
+        Map<String, String> summary = table.currentSnapshot().summary();
+        if (summary != null) {
+
+          if (summary.containsKey(SnapshotSummary.TOTAL_DATA_FILES_PROP)) {
+            stats.put(StatsSetupConst.NUM_FILES, summary.get(SnapshotSummary.TOTAL_DATA_FILES_PROP));
+          }
+
+          if (summary.containsKey(SnapshotSummary.TOTAL_RECORDS_PROP)) {
+            long totalRecords = Long.parseLong(summary.get(SnapshotSummary.TOTAL_RECORDS_PROP));
+            if (summary.containsKey(SnapshotSummary.TOTAL_EQ_DELETES_PROP) &&
```
[the remark this answers is not in the document]
Comment (simhadri-g) [reply iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergStorageHandler.java:351]: We still have to verify if the snapshot summary contains the TOTAL_EQ_DELETES_PROP key. In most cases, it should be there but if it is not, we want to fall back to total count.

[the remark this answers is not in the document]
Comment (deniskuzZ) [reply iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergStorageHandler.java:351]: 👍

**Review by deniskuzZ — APPROVED** [review]

**Comment by simhadri-g** [comment]
Thanks @deniskuzZ , @zhangbutao , @aturoczy , @InvisibleProgrammer  for the review !
